### PR TITLE
3D ResNet feature extraction

### PR DESCRIPTION
**Pull request by Henryy-rs**
Hello! I am a student studying computer vision.
Thanks for your great job.

I tried to extract feature from 3D ResNet. 
I used a pretrained model from here.

https://github.com/kenshohara/3D-ResNets-PyTorch
(which i used is.. r3d101_KM_200ep.pth: --model resnet --model_depth 101 --n_pretrain_classes 1039)

I got better results than C3D, I3D.

image

3D ResNet, 60k iterations, AUC=0.802

Could you accept my pull request?

### CODE REVIEW

**Comment by ekosman** [comment]
Thanks! I'll review this soon :)

File: TorchUtils.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
+import logging
```
Comment (ekosman) [inline TorchUtils.py:1]: This file seems like a duplicate of network/TorchUtils.
Isn't it?

File: feature_extractor.py [hunk 2]
```diff
@@ -132,7 +132,7 @@ def read_features(file_path, cache=None):
     features = None
     with open(file_path, 'r') as fp:
         data = fp.read().splitlines(keepends=False)
-        features = np.zeros((len(data), 4096))
+        features = np.zeros((len(data), 2048))
```
Comment (ekosman) [inline feature_extractor.py:135]: Should check the feature extractor type used in order to specify the right dimensions

File: TorchUtils.py (hunk 1 above)
Comment (ekosman) [inline TorchUtils.py:1]: This file seems like a duplicate of network/TorchUtils.
Isn't it?

File: network/anomaly_detector_model.py [hunk 3]
```diff
@@ -40,10 +40,10 @@ def custom_objective(y_pred, y_true):
     anomal_segments_scores = y_pred[anomal_vids_indices].squeeze(-1)  # (batch/2, 32, 1)
 
     # get the max score for each video
-    normal_segments_scores_maxes = normal_segments_scores.max(dim=-1)[0]
-    anomal_segments_scores_maxes = anomal_segments_scores.max(dim=-1)[0]
+    normal_segments_scores_maxes = normal_segments_scores.topk(k=2, dim=-1)[0] + normal_segments_scores.topk(k=2, largest=False, dim=-1)[0]
```
Comment (ekosman) [inline network/anomaly_detector_model.py]: The right term takes the minimal value. Why is that?

File: network/anomaly_detector_model.py [hunk 4]
```diff
@@ -40,10 +40,10 @@ def custom_objective(y_pred, y_true):
...
+    anomal_segments_scores_maxes = anomal_segments_scores.topk(k=2, dim=-1)[0] + normal_segments_scores.topk(k=2, largest=False, dim=-1)[0]
 
-    hinge_loss = 1 - anomal_segments_scores_maxes + normal_segments_scores_maxes
+    hinge_loss = 4 - torch.sum(anomal_segments_scores_maxes, dim=-1) + torch.sum(normal_segments_scores_maxes, dim=-1)
```
Comment (ekosman) [inline network/anomaly_detector_model.py]: This is not the loss function described in the original paper

**Review by ekosman — CHANGES_REQUESTED** [review]

File: network/anomaly_detector_model.py (hunk 4 above)
Comment (Henryy-rs) [reply network/anomaly_detector_model.py]: I tried to make several variations from the loss function in the original paper. It's one of them.
I fixed it with the loss function from the original paper in the latest commit!

File: feature_extractor.py (hunk 2 above)
Comment (Henryy-rs) [reply feature_extractor.py:135]: I think so too. I will fix it.

File: network/anomaly_detector_model.py (hunk 3 above)
Comment (Henryy-rs) [reply network/anomaly_detector_model.py]: I tried to make several variations from the loss function in the original paper. 
Those with I3D feature extraction worked(+0.02 AUC), but did not work when using 3D ResNet. So I used the loss function in the original paper to get the result.

File: TorchUtils.py (hunk 1 above)
Comment (Henryy-rs) [reply TorchUtils.py:1]: Yes. I forgot to remove it😂 I will fix it.

File: feature_extractor.py [hunk 5]
```diff
@@ -132,12 +132,13 @@ def read_features(file_path, cache=None):
...
         data = fp.read().splitlines(keepends=False)
-        features = np.zeros((len(data), 4096))
+        features = np.zeros((len(data), feature_dim))
         for i, line in enumerate(data):
             features[i, :] = [float(x) for x in line.split(' ')]
 
     features = torch.from_numpy(features).float()
-    cache[file_path] = features
+    if cache is not None:
+        cache[file_path] = features
```
Comment (Henryy-rs) [inline feature_extractor.py:141]: This because I commited as "change" and reset it.
I didn't know that you committed the name "change" before.

File: feature_extractor.py [hunk 6]
```diff
@@ -123,7 +123,7 @@ def write(self, feature, video_name, idx, dir):
         self.store(feature, idx)
 
 
-def read_features(file_path, cache=None):
+def read_features(file_path, feature_dim, cache=None):
```
Comment (Henryy-rs) [inline feature_extractor.py:126]: I added a parameter feature_dim. 
Which is better, passing feature dimension or feature extractor type?

Comment (ekosman) [reply feature_extractor.py:126]: This is good

**Review by ekosman — APPROVED** [review]

**Comment by Henryy-rs** [comment]
Thanks for approving my PR🤗

**Comment by dmascali** [comment]
Hi guys, great work!

I am trying to reproduce your AUC with resnet-extracted features. After 1100 epochs with 20k iterations each, I only get AUC = 0.77.  
Screenshot_2021-12-22_19-30-20 (2)



You said that you obtained 0.802 with 60k iterations? What do you mean exactly? Did you set iterations=20k and epochs=3, or did you set epoch to 60k and iterations to 20k? 

I am having a hard time understanding _iterations_ and _epochs_ in this implementation. In the original code from the Sultani it seems that they set iterations to 20k and trained for just one “epoch”. On the contrary, in this pytorch implementation the checkpoints names make me think that 20k epochs were run (each with 20k iterations? This would mean 400000k iterations in total!).
Despite a good GPU, I don’t think I can reach 20k epochs in a reasonable time (to get to 1k it took me almost 2 days).

Thank you in advance for any clarification

**Comment by Henryy-rs** [comment]
Hello dmascali,

Yes I set iteration=20k, epochs=3(lr=0.01)
In my opinion, 1100 epochs are too many will result in overfitting.

Also I saw the latest checkpoint on this repo,
I think it comes from old version of training code or other reasons.

**Comment by dmascali** [comment]
Thank you for the reply! 
It makes sense. 

Btw, I found the maximum AUC after 5 epochs (0.805).

roc_auc

**Comment by Henryy-rs** [comment]
Great!
I tried 3d resnet with depth=152 , got better results than 101. 

basic_roc_auc

**Comment by ekosman** [comment]
There's not much about epochs and iterations in this project as the data is randomly sampled in every iteration with replacement. It is just easier to manage model checkpoints that way.

The models on the master branch are from the detector trained on C3D features. I'll have to reorganize the directory to contain both detectors trained on C3D and Resnet. Please create PRs if you have trained models :)

**Comment by ekosman** [comment]
@Henryy-rs Also, if you have pre-computed features from the Resnet model, we can upload them here and share so that others can use it.

**Comment by Henryy-rs** [comment]
@ekosman 

This is my Google Drive link to precomputed ResNet features. 
Can you add this to readme?

https://drive.google.com/file/d/17wdy_DS9UY37J9XTV5XCLqxOFgXiv3ZK/view?usp=sharing

**Comment by ekosman** [comment]
@Henryy-rs 

I updated `README.md`.

**Comment by dmascali** [comment]
I added a PR with the trained model and the link to the extracted features. I used the first model proposed by Henryy-rs (depth 101). 

@Henryy-rs, Did you upload the features extracted with the deeper model (i.e., 152) ?

**Comment by Henryy-rs** [comment]
@dmascali
Yes, I uploaded the features from resnet  with depth 152.

**Comment by dmascali** [comment]
I updated the PR to include also the model trained with depth 152. Best AUC is at epoch 6.

**Comment by ekosman** [comment]
> I updated the PR to include also the model trained with depth 152. Best AUC is at epoch 6.

Thanks for the contribution!
